accepted/file-enumeration.md

```diff
@@ -0,0 +1,248 @@
+# Extensible File Enumeration
+
+**PM** [Immo Landwerth](https://github.com/terrajobst) | **Dev** [Jeremy Kuhne](https://github.com/jeremykuhne)
+
+Enumerating files in .NET provides limited configurability. You can specify a simple DOS style pattern and whether or not to look recursively. More complicated filtering requires post filtering all results which can introduce a significant performance drain.
+
+Recursive enumeration is also problematic in that there is no way to handle error states such as access issues or cycles created by links.
+
+These restrictions have a significant impact on file system intensive applications, a key example being MSBuild. This document proposes a new set of primitive file and directory traversal APIs that are optimized for providing more flexibility while keeping the overhead to a minimum so that enumeration becomes both more powerful as well as more performant.
+
+## Scenarios and User Experience
+
+1. MSBuild can custom filter filesystem entries with limited allocations and form the results in any desired format.
+2. Users can build custom enumerations utilizing completely custom or provided commonly used filters and transforms.
+
+## Requirements
+
+
+### Goals
+
+
+1. Custom filtering based on common file system data
+	- Name
+	- Attributes
+	- Time stamps
+    - File size
+2. Result transforms can be of any desired output type
+	- Like Linq Select(), but keeps FileData on the stack
+3. API minimizes allocations
+4. API is cross platform abstract
+3. We provide common filters and transforms
+	- To file/directory name
+	- To full path
+	- To File/Directory/FileSystemInfo
+	- DOS style filters (Legacy- `*/?` with DOS semantics, e.g. `*.` is all files without an extension)   
+	- Simple Regex filter
+	- Simpler globbing (`*/?` without DOS style variants)
+	- Set of extensions (`*.c`, `*.cpp`, `*.cc`, `*.cxx`, etc.)
+4. Recursive behavior is configurable
+	- On/Off
+	- Predicate based on FileData
+5. Can avoid throwing access denied exceptions
+
+### Non-Goals
```

Comment:
One non-goal to call out is that  we don't intend to replace the existing IO APIs -- these are meant to be advanced APIs for folks that really have to care about performance.